core/src/SYCL/Kokkos_SYCL_BlockSize_Deduction.hpp

```diff
@@ -0,0 +1,194 @@
+template <typename UnaryFunction, typename FunctorType, typename LaunchBounds, template <typename> class Wrapper>
+inline int sycl_deduce_block_size(bool early_termination,
+				  const sycl::queue& q,
+				  const FunctorType& f,
+                                  UnaryFunction block_size_to_dynamic_shmem,
+                                  LaunchBounds) {
+
+  // Get the device & check backend
+  const sycl::device sycl_device = q.get_device();
+  KOKKOS_ASSERT(sycl_device.get_backend() == sycl::backend::cuda);
+
+  // Get the compiled kernel to query register & memory usage
+  sycl::program p{q.get_context()};
+  p.build_with_kernel_type<Wrapper<FunctorType>>();
+  auto k = p.get_kernel<Wrapper<FunctorType>>();
+
+  auto num_regs = k.template get_info<
+      sycl::info::kernel_device_specific::ext_codeplay_num_regs>(sycl_device);
+
+  size_t kernelMaxThreadsPerBlock = k.template get_info<
+      sycl::info::kernel_device_specific::work_group_size>(sycl_device);
+
+#if defined(KOKKOS_ARCH_AMPERE80)
+  int const max_threads_per_sm = 2048;
+#else
+  // TODO other arches
+#endif
+
+  int const device_max_threads_per_block =
+    sycl_device.template get_info<sycl::info::device::max_work_group_size>();
+
+  int const max_threads_per_block =
+      std::min(LaunchBounds::maxTperB == 0 ? (int)device_max_threads_per_block
+                                           : (int)LaunchBounds::maxTperB,
+               (int)kernelMaxThreadsPerBlock);
+
+  int const min_blocks_per_sm =
+      LaunchBounds::minBperSM == 0 ? 1 : LaunchBounds::minBperSM;
+
+  // Recorded maximum
+  int opt_block_size     = 0;
+  int opt_threads_per_sm = 0;
+
+  size_t wg_size_multiple = k.template get_info<
+      sycl::info::kernel_device_specific::preferred_work_group_size_multiple>(sycl_device);
+
+  assert(max_threads_per_block % wg_size_multiple == 0);
+
+  for (int block_size = max_threads_per_block; block_size > 0;
+       block_size -= wg_size_multiple) {
+
+    // 'dynamic_shmem' is a misnomer here. It's allocated before launch by
+    // the host & it's sycl 'local' memory.
+    size_t const dynamic_shmem = block_size_to_dynamic_shmem(block_size);
+
+    int blocks_per_sm = sycl_max_active_blocks_per_sm(
+        block_size, dynamic_shmem, num_regs);
+
+    int threads_per_sm = blocks_per_sm * block_size;
+    if (threads_per_sm > max_threads_per_sm) {
+      blocks_per_sm  = max_threads_per_sm / block_size;
+      threads_per_sm = blocks_per_sm * block_size;
+    }
+
+    if (blocks_per_sm >= min_blocks_per_sm) {
+      if (threads_per_sm >= opt_threads_per_sm) {
+        opt_block_size     = block_size;
+        opt_threads_per_sm = threads_per_sm;
+      }
+    }
+
+    if (early_termination && opt_block_size != 0) break;
+  }
+
+  return opt_block_size;
+}
+
+template <class FunctorType, class LaunchBounds,
+          template <typename> class Wrapper>
+int sycl_get_opt_block_size(const sycl::queue& q, const FunctorType& f,
```

Comment:
I'm ignorant of our SYCL backend, here. Would it be possible for this to take in a Kokkos::SYCL instance (by cref)? That would help enable autotuning later

Reply:
That shouldn't be a problem.